compiler/lib/effects.ml

```diff
@@ -0,0 +1,436 @@
+        let allocate_continuation f =
+          let constr_cont, k' =
+            (* Construct continuation: it binds the return value [x],
+               allocates closures for dominated blocks and jumps to the
+               next block. *)
+            let pc, args = cont in
+            let ret = Var.fresh () in
+            let f' = closure_of_pc ~st pc in
+            allocate_closure
+              ~st
+              ~params:[ x ]
+              ~body:
+                (alloc_jump_closures @ [ Let (ret, Apply { f = f'; args; exact = true }) ])
+              ~branch:(Return ret)
+          in
+          let ret = Var.fresh () in
+          Some (body_prefix, constr_cont @ f ret k', Return ret)
+        in
+        match e with
+        | Apply { f; args; exact; _ } ->
+            allocate_continuation
+            @@ fun ret k' -> [ Let (ret, Apply { f; args = args @ [ k' ]; exact }) ]
+        | Prim (Extern "%resume", [ Pv stack; Pv f; Pv arg ]) ->
+            allocate_continuation
+            @@ fun ret k' ->
+            (* Resume the continuation [stack] and invoke function [f]
+               in this context *)
+            let k'' = Var.fresh_n "cont" in
+            [ Let (k'', Prim (Extern "caml_resume_stack", [ Pv stack; Pv k' ]))
+            ; Let (ret, Apply { f; args = [ arg; k'' ]; exact = false })
+            ]
+        | Prim (Extern "%perform", [ Pv effect ]) ->
+            allocate_continuation
+            @@ fun ret k' ->
+            [ Let
+                ( ret
+                , Prim (Extern "caml_perform_effect", [ Pv effect; Pc (Int 0l); Pv k' ])
+                )
+            ]
+        | _ -> None)
+    | None, _
+    | Some (_, (Set_field _ | Offset_ref _ | Array_set _ | Assign _)), _
+    | _, (Raise _ | Stop | Cond _ | Switch _ | Pushtrap _ | Poptrap _) -> None
+  in
+
+  let body, last =
+    match rewritten_block with
+    | Some (body_prefix, last_instrs, last) ->
+        List.map body_prefix ~f:(fun i -> cps_instr ~st i) @ last_instrs, last
+    | None ->
+        let last_instrs, last = cps_last ~st block.branch ~k in
+        let body =
+          List.map block.body ~f:(fun i -> cps_instr ~st i)
+          @ alloc_jump_closures
+          @ last_instrs
+        in
+        body, last
+  in
+
+  { params = block.params; body; branch = last }
+
+let split_blocks (p : Code.program) =
+  (* Ensure that function applications and effect primitives are in
+     tail position *)
+  let split_block pc block p =
+    let is_split_point i r branch =
+      match i with
+      | Let (x, (Apply _ | Prim (Extern ("%resume" | "%perform"), _))) -> (
+          (not (List.is_empty r))
+          ||
+          match branch with
+          | Branch _ -> false
+          | Return x' -> (
+              (not (Var.equal x x'))
+              ||
+              match i with
+              | Let (_, Prim (Extern "%perform", _)) -> true
+              | _ -> false)
+          | _ -> true)
```

Comment:
I don't follow lines 334 and 336. Could you add a comment ?

Reply:
`cps_block` was not rewriting `%perform` when followed by a return. I have added a case in `cps_block` to handle this case and removed this code.